src/mesh_processor/src/generate_mesh.cpp

```diff
@@ -0,0 +1,160 @@
+#include "generate_mesh.h"
+
+#include <boost/range/algorithm.hpp>
+#include <cinolib/find_intersections.h>
+#include <cinolib/profiler.h>
+#include <cinolib/tetgen_wrap.h>
+
+#include <execution>
+
+#include "utils.h"
+
+namespace {
+bool operator>(const cinolib::vec3d& lhs, const cinolib::vec3d& rhs) {
+	return std::tie(lhs.x(), lhs.y(), lhs.z()) >
+	       std::tie(rhs.x(), rhs.y(), rhs.z());
+}
+}  // namespace
+
+namespace MeshProcessor {
+using namespace cinolib;
+
+CustomTetmesh MeshGenerator::get_tet_mesh() {
+	Profiler profiler;
+	profiler.push("mesh creation");
+
+	auto tet_meshes = generate_layers_meshes(representation_.layers());
+	auto heat_data  = representation_.shapes_metadata();
+	assert(heat_data.size() == tet_meshes.size() &&
+	       "Heat data and shapes mismatch");
+	CustomTetmesh total_tetmesh;
+	for (auto i = 0; i < tet_meshes.size(); ++i) {
+		const auto shape_heat_data = heat_data[i];
+		auto& tet_mesh             = tet_meshes[i];
+		auto mesh_volume           = tet_mesh.mesh_volume();
+		auto& polys                = tet_mesh.vector_polys();
+		for (auto pid = 0; pid < polys.size(); ++pid) {
+			auto& p_data                = tet_mesh.poly_data(pid);
+			p_data.thermal_conductivity = shape_heat_data.thermal_conductivity;
+			p_data.ambient_temperature  = shape_heat_data.ambient_temperature;
+			p_data.heat_flow            = shape_heat_data.heat_flow;
+			p_data.intensity_of_heat_source =
+					shape_heat_data.power * tet_mesh.poly_volume(pid) / mesh_volume;
+			p_data.convective_heat = shape_heat_data.convective_heat;
+		}
+		total_tetmesh += tet_mesh;
```

Comment:
please avoid multiple data duplication, you can store just shape_id instead. 
(in local example data pack increases in 50 times)

Reply:
let it be an improvement for future

Reply:
not part of this review, lets make mvp